File: 1603_parking_system/parking_system.cpp
/*
 * @Date: 2021-03-19 08:21:22
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-03-19 08:24:35
 */

#include <cassert>

class ParkingSystem {
 public:
  ParkingSystem(int big, int medium, int small) : park_{0, 0, 0} {
    park_[0] = big;
    park_[1] = medium;
    park_[2] = small;
  }

  bool addCar(int carType) {
    if (park_[carType - 1] == 0)
      return false;
    else
      --park_[carType - 1];
    return true;
  }

 private:
  // vector<int> park_;
  int park_[3];
};

int main() {
  ParkingSystem* obj = new ParkingSystem(1, 1, 0);
  assert(obj->addCar(1));
  assert(obj->addCar(2));
  assert(!obj->addCar(3));
  assert(!obj->addCar(1));
  return 0;
}

File: rust/src/1603_parking_system/parking_system.rs
/*
 * @Date: 2021-03-19 08:21:30
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-03-19 08:26:19
 */

struct ParkingSystem([i16; 3]);

/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl ParkingSystem {
    #[inline(always)]
    fn new(big: i32, medium: i32, small: i32) -> Self {
        Self([big as i16, medium as i16, small as i16])
    }
    #[inline(always)]
    fn add_car(&mut self, car_type: i32) -> bool {
        if self.0[car_type as usize - 1] == 0 {
            return false;
        } else {
            self.0[car_type as usize - 1] -= 1;
        }
        true
    }
}

fn main() {
    let mut obj = ParkingSystem::new(1, 1, 0);
    assert!(obj.add_car(1));
    assert!(obj.add_car(2));
    assert!(!obj.add_car(3));
    assert!(!obj.add_car(1));
}
